### PR TITLE
reflect: stable type path v2

### PR DESCRIPTION
**Pull request by soqb**
# Objective

- Introduce a stable alternative to [`std::any::type_name`](https://doc.rust-lang.org/std/any/fn.type_name.html).
- Rewrite of #5805 with heavy inspiration in design.
- On the path to #5830.
- Part of solving #3327.

## Solution

- Add a `TypePath` trait for static stable type path/name information.
- Add a `TypePath` derive macro.
- Add a `impl_type_path` macro for implementing internal and foreign types in `bevy_reflect`.

---

## Changelog

- Added `TypePath` trait.
- Added `DynamicTypePath` trait and `get_type_path` method to `Reflect`.
- Added a `TypePath` derive macro.
- Added a `bevy_reflect::impl_type_path` for implementing `TypePath` on internal and foreign types in `bevy_reflect`.
- Changed `bevy_reflect::utility::(Non)GenericTypeInfoCell` to `(Non)GenericTypedCell<T>` which allows us to be generic over both `TypeInfo` and `TypePath`.
- `TypePath` is now a supertrait of `Asset`, `Material` and `Material2d`.
- `impl_reflect_struct` needs a `#[type_path = "..."]` attribute to be specified.
- `impl_reflect_value` needs to either specify path starting with a double colon (`::core::option::Option`) or an `in my_crate::foo` declaration.
- Added `bevy_reflect_derive::ReflectTypePath`.
- Most uses of `Ident` in `bevy_reflect_derive` changed to use `ReflectTypePath`.

## Migration Guide

- Implementors of `Asset`, `Material` and `Material2d` now also need to derive `TypePath`.
- Manual implementors of `Reflect` will need to implement the new `get_type_path` method.

## Open Questions
- [x] ~This PR currently does not migrate any usages of `std::any::type_name` to use `bevy_reflect::TypePath` to ease the review process. Should it?~ Migration will be left to a follow-up PR.
- [ ] This PR adds a lot of `#[derive(TypePath)]` and `T: TypePath` to satisfy new bounds, mostly when deriving `TypeUuid`. Should we make `TypePath` a supertrait of `TypeUuid`? [Should we remove `TypeUuid` in favour of `TypePath`?](https://github.com/bevyengine/bevy/pull/5805/files/2afbd855327c4b68e0a6b6f03118f289988441a4#r961067892)

### CODE REVIEW

File: crates/bevy_reflect/src/impls/std.rs
Comment (soqb) [inline crates/bevy_reflect/src/impls/std.rs:580]: is `bevy_utils::HashMap` the best path for this type? should we use `hashbrown::HashMap` instead?

**Comment by alice-i-cecile** [comment]
>  This PR currently does not migrate any usages of std::any::type_name to use bevy_reflect::TypePath to ease the review process. Should it?

I'd prefer to do this in a separate PR: this is already very large.

File: crates/bevy_reflect/src/impls/std.rs
Comment (alice-i-cecile) [reply crates/bevy_reflect/src/impls/std.rs:580]: bevy_util's HashMap is just a re-export of hashbrown's version IIRC.

File: crates/bevy_reflect/src/type_path.rs
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/type_path.rs:1]: Needs top level doc strings explaining why it exists, what it's used for, and why not just use `TypeId`.

File: crates/bevy_reflect/src/impls/std.rs
Comment (soqb) [reply crates/bevy_reflect/src/impls/std.rs:599]: it is. i can swing either way on `HashMap` - but i think i prefer `bevy_utils::HashMap` in case we want to switch to a different/our own `HashMap` implementation.

there is also the case of `bevy_utils::Instant` which currently fails builds on wasm.
we use the `instant` crate which includes its own implementation on wasm, since `std::time::Instant` is only available with wasi.

for now i will go with `bevy_utils::Instant` and `bevy_utils::HashMap` for future-proofing's sake.

**Comment by MrGVSV** [comment]
> Alternative to #5830.

For reference, #5830 is meant to work in tandem with this PR. Basically, there may only ever be one stable type name for a type, but there can be zero or more aliases (including ones marked "deprecated"). So they're not truly at odds with each other (though, whether or not we want actually want aliases is another story haha)

**Comment by MrGVSV** [comment]
> * This PR adds a lot of `#[derive(TypePath)]` and `T: TypePath` to satisfy new bounds, mostly when deriving `TypeUuid`. Should we make `TypePath` a supertrait of `TypeUuid`? [Should we remove `TypeUuid` in favour of `TypePath`?](https://github.com/bevyengine/bevy/pull/5805/files/2afbd855327c4b68e0a6b6f03118f289988441a4#r961067892)

My instinct says we should remove it. They essentially do the same thing; however, `TypeUuid` is `const`, which may be a determining factor. Either way, I believe this question might be better explored in a followup issue/PR as well.

**Comment by soqb** [comment]
> we want to implement Reflect for types we don't control, namely std and glam (are there more?)

yes there are a few more: `smallvec` of course, a couple in `bevy_utils` that currently link to external crates (`hashbrown::HashMap` and `instant::Duration`) but for future-proofing have paths in `bevy_utils` and `bevy_math::Rect`.

> we want to rely less on std::any::type_name for non-debug purposes because it is not its intended use

yep.

> I think the biggest issue with just using typename aliases was that Vec<Vec3> also needs to be stable, and when treating std::any::type_name as an opaque string (which we should) then we can't really make that work...

another potential risk with relying on `std::any::type_name` is that it may differ with different compiler flags, e.g. more verbose with `debug = true`. solving this without a dedicated `TypePath` trait would require either annotating every single `Reflect` with an alias, or accepting that all scenes will have to be built twice.

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:478]: Nit/Bikeshed: I think this should be named `ReflectTypePath` (or something similar). `PathToType` is lexically confusing since `To` gives this enum three possible interpretations when reading:
1. This contains the path for a given type
2. This contains a conversion from a path to a type
3. This contains a path that requires typing

`ReflectTypePath` avoids this possible confusion and is more consistent with most of the other types in this module. It's also what contains the data necessary to implement `TypePath`, much like `ReflectStruct` contains the data necessary to implement `Struct`.

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:484]: Nit: Type aliases (#5830) may or may not be merged in at some point in the future. _If_ it is, using `alias` here may be confusing. Perhaps we can use something more like `custom_path`? (And `is_aliased` can become `is_custom`.)

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:509]: A lot of these methods are prefixed with `named_`. Are these meant to read `name`? Otherwise, they all sound like they should return `bool` ("is this named as an ident?").

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:641]: Nit: What is a `name`? Here it seems like it's a stingified ident. Would it be more clear to name this `ident_string` or `ident_as_lit_str`?

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:493]: Nit: Why call this `name` and not `ident`?

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:477]: Nit: I feel like this needs a bit more internal documentation.

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:559]: Nit: Maybe I haven't gotten to that part yet, but what do you mean when you say `` Returns an expression for an `AsRef<str>` ``?

The docs on these methods could give a little more info (or at least an example since many return opaque `TokenStream`s).

**Review by MrGVSV — COMMENTED** [review]
Just starting review. I've only looked at pretty much one file for now so take these comments with a grain of salt 😅

Most of them are mainly related to naming, documentation, and code clarity (specifically for future contributors' benefit).

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (soqb) [inline crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:473]: this change is actually not necessary with the introduction of `parameter_{types,trait_bounds}`

File: crates/bevy_reflect/src/reflect.rs
Comment (PROMETHIA-27) [inline crates/bevy_reflect/src/reflect.rs:241]: This impl seems suspect and prone to causing mistakes? Why is it necessary?

**Review by PROMETHIA-27 — APPROVED** [review]
I'd like to get some momentum on this again; we really should not be relying on `std::any::type_name()`.

File: crates/bevy_reflect/src/enums/mod.rs
Comment (MrGVSV) [inline crates/bevy_reflect/src/enums/mod.rs:319]: Nit: This isn't needed anymore since we add the bounds, right?
```suggestion
        enum TestEnum<T: FromReflect> {
```

File: examples/reflection/generic_reflection.rs
Comment (MrGVSV) [inline examples/reflection/generic_reflection.rs:16]: Nit: Same here. This isn't needed anymore, right?
```suggestion
struct MyType<T: Reflect> {
```

**Review by MrGVSV — CHANGES_REQUESTED** [review]
This looks great!

Just some comments and nitpicks, but this is looking pretty much ready to go soon!

File: crates/bevy_reflect/src/reflect.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/reflect.rs:241]: I agree, I think we probably want to hold off on this if we can (unless it was in fact necessary for something).

**Comment by cart** [comment]
> This PR adds a lot of #[derive(TypePath)] and T: TypePath to satisfy new bounds, mostly when deriving TypeUuid. Should we make TypePath a supertrait of TypeUuid? [Should we remove TypeUuid in favour of TypePath?](https://github.com/bevyengine/bevy/pull/5805/files/2afbd855327c4b68e0a6b6f03118f289988441a4#r961067892)

Haven't done a review yet, but this is on my mind currently for the asset system rework (ive removed the TypeUuid requirement for Assets and rely on type names currently, with the intent to switch to type paths once they land).

I still think we should remove TypeUuid in favor of stable type paths. The UX of type uuids isn't great and names are much easier to understand / debug. Type paths are good enough for components, they should be good enough for everything else.

**Comment by MrGVSV** [comment]
> I still think we should remove TypeUuid in favor of stable type paths. The UX of type uuids isn't great and names are much easier to understand / debug. Type paths are good enough for components, they should be good enough for everything else.

I agree, but this should probably happen in a followup PR, right? There may be unforeseen issues with replacing `TypeUuid` or questions that might arise. It's probably better to keep that in its own PR so this one can be purely focused on _adding_ `TypePath`.

File: crates/bevy_reflect/bevy_reflect_derive/src/type_path.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/type_path.rs]: Gotcha, okay that makes sense. Ideally, we'd have a compile-fail test for this, but I don't think it's a huge issue if we don't since it's more internal-facing.

We should still think about wording here to avoid the term `alias` and maybe make this error case a bit clearer. Maybe like ``type paths must be fully-qualified and start with a double-colon (`::`))``?

**Comment by soqb** [comment]
definitely should be in a follow up PR, this PR doesn't even migrate any uses of `std::any::type_name` to avoid making the diff even bigger.

**Comment by cart** [comment]
> I agree, but this should probably happen in a followup PR, right? There may be unforeseen issues with replacing TypeUuid or questions that might arise. It's probably better to keep that in its own PR so this one can be purely focused on adding TypePath.

Fully agreed on that. Just answering the "open question" from my perspective. I definitely don't think we should make TypePath a supertrait of TypeUuid

File: crates/bevy_reflect/src/type_path.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/type_path.rs]: Resolved

File: crates/bevy_reflect/src/type_path.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/type_path.rs]: Resolved

File: crates/bevy_reflect/src/type_path.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/type_path.rs]: Resolved

File: crates/bevy_reflect/src/impls/std.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/impls/std.rs:1336]: After reading the newly added section, I'm still curious if this would count as an "anonymous" type. Obviously, `Path` itself isn't, but I feel like the `&` makes it so. Is my thinking here correct?

File: crates/bevy_reflect/bevy_reflect_derive/src/type_path.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/type_path.rs]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/type_path.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/type_path.rs]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/type_path.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/type_path.rs]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:206]: Well, we've updated to `syn2`. But again, this can be addressed in a future PR where we reconsider exactly what kind of syntax we want for attributes.

File: crates/bevy_reflect/src/lib.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/lib.rs:1106]: New docs on the macro clear this up. Resolved.

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/type_path.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/type_path.rs]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:214]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:135]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:321]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:439]: Resolved

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:496]: Resolved. Great examples!

File: crates/bevy_reflect/bevy_reflect_derive/src/utility.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/utility.rs:318]: Not really necessary. We can consider this resolved as well

File: crates/bevy_reflect/bevy_reflect_derive/src/utility.rs
Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/utility.rs]: Resolved

File: crates/bevy_reflect/src/enums/mod.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/enums/mod.rs]: Resolved

File: examples/reflection/generic_reflection.rs
Comment (MrGVSV) [reply examples/reflection/generic_reflection.rs]: Resolved

**Review by MrGVSV — APPROVED** [review]
Looks good! I think I have one or two unresolved review comments left, but once those are addressed this should be ready to go!

File: crates/bevy_reflect/src/impls/std.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/impls/std.rs:1336]: Oh wait. I just realized that the fact that this doesn't implement any of the other methods directly implies that it's "anonymous". My bad! Carry on!

File: crates/bevy_macros_compile_fail_tests/tests/deref_mut_derive/missing_deref.fail.stderr
Comment (devil-ira) [inline crates/bevy_macros_compile_fail_tests/tests/deref_mut_derive/missing_deref.fail.stderr:1]: For some reason these updated error messages appear locally but not in CI. (Noticed in #8738)
Reverting this file should let the check-compiles part of CI pass

**Comment by MrGVSV** [comment]
Going to remove the `S-Controversial` tag for now since this seems to be the way we want to go. If there are any concerns raised, we can re-tag it.

File: crates/bevy_reflect/src/reflect.rs
Comment (cart) [inline crates/bevy_reflect/src/reflect.rs:96]: Why aren't we just doing `Reflect: DynamicTypePath`? Seems like this just makes the trait more complicated / does more codegen for something that rust can handle for us. Was there rationale here?

**Review by cart — COMMENTED** [review]
Did a "post merge review" and this looks good to me. Just one question / comment.

File: crates/bevy_reflect/src/reflect.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/reflect.rs:96]: I think the reason was to match `TypeInfo`. But adding it as a supertrait probably makes things a bit simpler.

Comment (MrGVSV) [reply crates/bevy_reflect/src/reflect.rs:96]: Another thought is that we may want to have it eventually return the _represented_ type path. In which case, it will need to be changed to return an `Option`.

Comment (cart) [reply crates/bevy_reflect/src/reflect.rs:96]: Yeah I think thats the move. I've got a PR prepped.

**Pull request merged**
